samcli/lib/iac/cfn/cfn_iac.py

```diff
@@ -1,33 +1,172 @@
+        for key, value in template_dict.items():
+            stack[key] = value
+
+        resources_section = stack.get("Resources", DictSection())
+        for resource in resources_section.section_items:
+            resource_id = resource.item_id
+            resource_type = resource.get("Type", None)
+            properties = resource.get("Properties", {})
+            package_type = properties.get("PackageType", ZIP)
+
+            resource_assets = []
+
+            if resource_type in NESTED_STACKS_RESOURCES:
+                nested_stack = self._extract_nested_stack(path, resource_id, properties, resource_type)
+                resource.nested_stack = nested_stack
+
+            if resource_type in RESOURCES_WITH_LOCAL_PATHS:
+                for path_prop_name in RESOURCES_WITH_LOCAL_PATHS[resource_type]:
+                    asset_path = jmespath.search(path_prop_name, properties)
+                    if is_local_path(asset_path) and package_type == ZIP:
+                        reference_path = base_dir if resource_type in BASE_DIR_RESOURCES else os.path.dirname(path)
+                        asset_path = get_local_path(asset_path, reference_path)
+                        asset = S3Asset(source_path=asset_path, source_property=path_prop_name)
+                        resource_assets.append(asset)
+                        stack.assets.append(asset)
+
+            if resource_type in RESOURCES_WITH_IMAGE_COMPONENT:
+                for path_prop_name in RESOURCES_WITH_IMAGE_COMPONENT[resource_type]:
+                    asset_path = jmespath.search(path_prop_name, properties)
+                    if asset_path and package_type == IMAGE:
+                        asset = ImageAsset(source_local_image=asset_path, source_property=path_prop_name)
+                        resource_assets.append(asset)
+                        stack.assets.append(asset)
+
+            resource.assets = resource_assets
+
+        metadata_section = stack.get("Metadata", DictSection())
+        for metadata in metadata_section.section_items:
+            if not isinstance(metadata, DictSectionItem):
+                continue
+            metadata_type = metadata.item_id
+            metadata_body = metadata.body
+            metadata_assets = []
+            if metadata_type in METADATA_WITH_LOCAL_PATHS:
+                for path_prop_name in METADATA_WITH_LOCAL_PATHS[metadata_type]:
+                    asset_path = jmespath.search(path_prop_name, metadata_body)
+                    asset = S3Asset(source_path=asset_path, source_property=path_prop_name)
+                    metadata_assets.append(asset)
+                    stack.assets.append(asset)
+
+            metadata.assets = metadata_assets
+
+        stack.extra_details[TEMPLATE_PATH_KEY] = path
+        return stack
+
+    def _extract_nested_stack(
+        self, parent_stack_template_path, resource_id, properties, resource_type
+    ) -> Optional[Stack]:
```

Comment:
updated to return `Optional[Stack]` instead of `Stack`. Fixes typing issues since, this can return `None`